wave_lang/kernel/wave/mlir_converter.py

```
@@ -0,0 +1,289 @@
+            "name": getattr(node, "name", ""),
+            "op": getattr(node, "op", ""),
+            "tkw_op_name": getattr(custom, "tkw_op_name", "unknown"),
+            "subgraph": getattr(node, "subgraph_name", ""),
+        }
+        # Capture type info for placeholders.
+        if isinstance(custom, Placeholder):
+            t = getattr(custom, "_type", None)
+            if t is not None:
+                if issubclass(t, Memory):
+                    shape_syms = _shape_syms(t)
+                    addr = _addr_str(
+                        getattr(t, "address_space", AddressSpace.GLOBAL_MEMORY)
+                    )
+                    dt = _dtype_str(getattr(t, "dtype", None))
+                    entry["wave_type"] = {
+                        "kind": "memory",
+                        "shape": shape_syms,
+                        "address_space": addr,
+                        "dtype": dt,
+                    }
+                elif issubclass(t, Register):
+                    shape_syms = _shape_syms(t)
+                    dt = _dtype_str(getattr(t, "dtype", None))
+                    entry["wave_type"] = {
+                        "kind": "register",
+                        "shape": shape_syms,
+                        "address_space": "register",
+                        "dtype": dt,
+                    }
+        # Capture result type for other ops when available on custom.
+        t = getattr(custom, "type", None)
+        if isinstance(t, type):
+            if issubclass(t, Register):
+                shape_syms = _shape_syms(t)
+                dt = _dtype_str(getattr(t, "dtype", None))
+                entry["wave_type"] = {
+                    "kind": "register",
+                    "shape": shape_syms,
+                    "address_space": "register",
+                    "dtype": dt,
+                }
+            elif issubclass(t, Memory):
+                shape_syms = _shape_syms(t)
```

```suggestion
            else:
                assert issubclass(t, Memory)
                shape_syms = _shape_syms(t)
```

---

Let's not just silently ignore unhandled values.